src/cpu/testers/traffic_gen/gups_gen.cc

```diff
@@ -0,0 +1,321 @@
+    requestorId(system->getRequestorId(this)),
+    nextGenEvent([this]{ generateNextReq(); }, name()),
+    nextSendEvent([this]{ sendNextBatch(); }, name()),
+    port(name() + ".port", this),
+    startAddr(params.start_addr),
+    memSize(params.mem_size),
+    blockSize(params.block_size),
+    doneReading(false),
+    onTheFlyRequests(0)
+{}
+
+Port&
+GUPSGen::getPort(const std::string &if_name, PortID idx)
+{
+    if (if_name != "port"){
+        return ClockedObject::getPort(if_name, idx);
+    }
+    else{
+        return port;
+    }
+}
+
+void
+GUPSGen::startup()
+{
+
+    uint64_t stride_size = blockSize / sizeof(uint64_t);
+
+    for (uint64_t start_index = 0; start_index < tableSize;
+                                start_index += stride_size)
+    {
+        uint8_t write_data[blockSize];
+        for (uint64_t offset = 0; offset < stride_size; offset++)
+        {
+            uint64_t value = start_index + offset;
+            std::memcpy(write_data + offset * sizeof(uint64_t),
+                        &value, sizeof(uint64_t));
+        }
+        Addr addr = indexToAddr(start_index);
+        PacketPtr pkt = getWritePacket(addr, blockSize, write_data);
+        port.sendFunctionalPacket(pkt);
+    }
+
+    /*
+     * initialize memory by functional access
+     * call generateNextBatch here.
+    */
+   schedule(nextGenEvent, curTick());
+}
+
+void
+GUPSGen::init()
+{
+
+    tableSize = memSize / sizeof(uint64_t);
+    numUpdates = 4 * tableSize;
+
+    for (int i = 0; i < 128; i++)
+    {
+        randomSeeds[i] = randomNumberGenerator((numUpdates / 128) * i);
+    }
+
+}
+
+void
+GUPSGen::generateNextReq()
+{
+    static uint64_t iterations = 0;
+    static int rIndex = 0;
+
+    assert (rIndex < 128);
+
+    if (iterations >= numUpdates / 128){
+        doneReading = true;
+
+        // exitSimLoop(name() + " has encountered the exit state and will "
+                // "terminate the simulation.\n");
+        return;
+    }
+
+    randomSeeds[rIndex] = (randomSeeds[rIndex] << 1) ^
+                    ((int64_t) randomSeeds[rIndex] < 0 ? POLY : 0);
+    uint64_t value = randomSeeds[rIndex];
+    uint64_t index = randomSeeds[rIndex] & (tableSize - 1);
+    Addr addr = indexToAddr(index);
+    updateTable[addr] = value;
+    PacketPtr pkt = getReadPacket(addr, sizeof(uint64_t));
+    DPRINTF(GUPSGen, "generateNextBatch: Pushed pkt, pkt->addr_range: %s,"
+                    "(pkt != nullptr): %d\n",
+                    pkt->getAddrRange().to_string(), pkt != nullptr);
+    requestPool.push(pkt);
+
+    rIndex++;
+    if (rIndex >= 128){
+        rIndex = 0;
+        iterations++;
+    }
+
+    schedule(nextGenEvent, curTick() + 250);
+
+    if (!nextSendEvent.scheduled() && !port.blocked()
+                            && !requestPool.empty()) {
+        schedule(nextSendEvent, curTick() + 100);
+    }
+}
+
+
+void
+GUPSGen::sendNextBatch()
+{
+    if (port.blocked()){
+        return;
+    }
+
+    if (!requestPool.empty())
```

Comment:
Why do you send things in "batches" and what is this "requestPool"?

Reply:
Sorry, the name is misleading, I'm not doing it in batches anymore, but I forgot to change the name of the function. requestPool is like a queue for the pending requests.